lib/rust/parser/src/macros/pattern.rs

```diff
@@ -341,6 +357,23 @@ impl Pattern {
                         Err(input)
                     },
             },
+            PatternData::Block => match input.pop_front() {
+                None => Err(default()),
+                Some(syntax::Item::Block(lines)) =>
+                    Ok(MatchResult::new(Match::Block(lines), input)),
+                Some(t) => {
+                    input.push_front(t);
+                    Err(input)
+                }
+            },
+            PatternData::NotBlock => match input.pop_front() {
+                None => Err(default()),
+                Some(t @ syntax::Item::Block(_)) => {
+                    input.push_front(t);
```

Comment:
Is this a kind of _"optimistic consumption"_ that assumes the input will be mostly valid? First of all `.pop_front()` and then on `Err(...)` returning the item back with `push_front`. Is it fast or would `match input.front() {` and only then `Match::NotBlock(input.pop_front())` be better in some cases?